Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -884,54 +894,67 @@ extension NonBlockingFileIO {
-    /// comprising of the whole file. The caller must close the returned `NIOFileHandle` when it's no longer needed.
-    ///
-    /// - note: The reason this returns the `NIOFileHandle` and the `FileRegion` is that both the opening of a file as well as the querying of its size are blocking.
+    /// The open file operation runs on a private thread pool which is separate from Task pool.
+    ///    
+    /// - note: The reason this provides the `NIOFileHandle` and the `FileRegion` is that both the opening of a file as well as 
+    /// the querying of its size are blocking.
     ///
     /// - parameters:
     ///     - path: The path of the file to be opened for reading.
-    /// - returns: the `NIOFileHandle` and the `FileRegion` comprising the whole file.
-    public func openFile(path: String) async throws -> (NIOFileHandle, FileRegion) {
-        let result = try await self.threadPool.runIfActive {
+    ///     - body: operation to run with file handle and region
+    /// - returns: return value of operation
+    public func withOpenFile<Result>(path: String, _ body: (NIOFileHandle, FileRegion) async throws -> Result) async throws -> Result {
+        let file = try await self.threadPool.runIfActive {
             let fh = try NIOFileHandle(path: path)
             do {
                 let fr = try FileRegion(fileHandle: fh)
-                return UnsafeTransfer((fh, fr))
+                return UnsafeTransfer((handle: fh, region: fr))
             } catch {
                 _ = try? fh.close()
                 throw error
             }
         }
-        return result.wrappedValue
+        defer {
+            try? file.wrappedValue.handle.close()
+        }
+        return try await body(file.wrappedValue.handle, file.wrappedValue.region)
     }
 
-    /// Open the file at `path` with specified access mode and POSIX flags on a private thread pool which is separate from any `EventLoop` thread.
+    /// Open file at `path` on a private thread pool, run an operation given the file handle and region and then close the file handle.
     ///
-    /// This function will return (a future) of the `NIOFileHandle` associated with the file opened.
-    /// The caller must close the returned `NIOFileHandle` when it's no longer needed.
+    /// The open file operation runs on a private thread pool which is separate from Task pool.
+    ///    
+    /// This function will return the result of the operation.
     ///
     /// - parameters:
     ///     - path: The path of the file to be opened for writing.
     ///     - mode: File access mode.
     ///     - flags: Additional POSIX flags.
     /// - returns: NIOFileHandle`.
-    public func openFile(path: String, mode: NIOFileHandle.Mode, flags: NIOFileHandle.Flags = .default) async throws -> NIOFileHandle {
-        let result = try await self.threadPool.runIfActive {
+    public func withOpenFile<Result>(
```

Comment:
Bikeshedding... I wonder if we need the "Open" here? It seems somewhat redundant.

Reply:
`withFile` doesn't make much sense to me, maybe `withFileHandle`? And should we keep the other function named `withOpenFile`